contracts/TopShotLocking.cdc

```diff
@@ -0,0 +1,135 @@
+import NonFungibleToken from 0xNFTADDRESS
+
+pub contract TopShotLocking {
+
+    // -----------------------------------------------------------------------
+    // TopShotLocking contract Events
+    // -----------------------------------------------------------------------
+
+    // Emitted when a Moment is locked
+    pub event MomentLocked(id: UInt64, duration: UFix64, expiryTimestamp: UFix64)
+
+    // Emitted when a Moment is unlocked
+    pub event MomentUnlocked(id: UInt64)
+
+    // Dictionary of locked NFTs
+    // nft resource uuid is the key
+    // locked until timestamp is the value
+    access(self) var lockedNFTs: {UInt64: UFix64}
+
+    // Dictionary of NFTs overridden to be unlocked
+    access(self) var unlockableNFTs: {UInt64: Bool} // nft resource uuid is the key
+
+    // isLocked Returns a boolean indicating if an nft exists in the lockedNFTs dictionary
+    //
+    // Parameters: nftRef: A reference to the NFT resource
+    //
+    // Returns: true if NFT is locked
+    pub fun isLocked(nftRef: &NonFungibleToken.NFT): Bool {
+        return self.lockedNFTs.containsKey(nftRef.uuid)
+    }
+
+    // getLockExpiry Returns the unix timestamp when an nft is unlockable
+    //
+    // Parameters: nftRef: A reference to the NFT resource
+    //
+    // Returns: unix timestamp
+    pub fun getLockExpiry(nftRef: &NonFungibleToken.NFT): UFix64 {
+        if !self.lockedNFTs.containsKey(nftRef.uuid) {
+            panic("NFT is not locked")
+        }
+        return self.lockedNFTs[nftRef.uuid]!
+    }
+
+    // lockNFT Takes an NFT resource and adds its unique identifier to the lockedNFTs dictionary
+    //
+    // Parameters: nft: NFT resource
+    //             expiryTimestamp: The unix timestamp in seconds after which the nft may be unlocked
+    //
+    // Returns: the NFT resource
+    pub fun lockNFT(nft: @NonFungibleToken.NFT, duration: UFix64): @NonFungibleToken.NFT {
+        let TopShotNFTType: Type = CompositeType("A.TOPSHOTADDRESS.TopShot.NFT")!
```

Comment:
i don't really like the idea of having template variables in the contract code

Reply:
have you considered having this contract setup in each user's account and the nftType can be passed to the function. This way it can be reusable by other SC. Then `lockedNFTs` is of type `{Type:{UInt64: UFix64}}`

Reply:
This relates to a discussion @joshuahannan and I had where we wanted to prevent someone from "spamming" the contract with non-topshot nfts and potentially slowing down the contract as the dictionary fills up.

Also, given that there is an admin backdoor which can unlock nfts prematurely, it makes the most sense to scope this purely for TopShot NFTs.

Reply:
i see. Good point